src/Pact/Types/Runtime.hs

```diff
@@ -434,17 +438,36 @@ throwArgsError FunApp {..} args s = throwErr ArgsError _faInfo $
   pretty s <> ", received " <> bracketsSep (map pretty args) <> " for " <>
             prettyFunTypes _faTypes
 
+throwOnChainArgsError :: Pretty n => FunApp -> [Term n] -> Eval e a
+throwOnChainArgsError FunApp{..} args = throwErr ArgsError _faInfo $
+  "Invalid arguments in call to"
+    <+> pretty _faName
+    <> ", received arguments of type "
+    <> bracketsSep (map (pretty . typeof') args) <> ", expected "
+    <> prettyFunTypes _faTypes
+
 throwErr :: PactErrorType -> Info -> Doc -> Eval e a
-throwErr ctor i err = get >>= \s -> throwM (PactError ctor i (_evalCallStack s) err)
+throwErr ctor i err = do
+  s <- use evalCallStack
+  inReplOrPreFork <- isInReplForkedError
+  throwM (PactError ctor i (if inReplOrPreFork then s else []) err)
 
 evalError :: Info -> Doc -> Eval e a
-evalError i = throwErr EvalError i
+evalError = throwErr EvalError
 
 evalError' :: HasInfo i => i -> Doc -> Eval e a
 evalError' = evalError . getInfo
 
+-- | Function to determine whether we are either pre-errors fork
+-- or in a repl environment.
+isInReplForkedError :: Eval e Bool
```

Comment:
`Bool` is only sort of convenient here, but I think I may prefer a data type like"

```
data ErrState = OnChain | OffChain
```
or something like this instead of `Bool`. It's just a convenient callsite. Major boolean blindness here

Reply:
I completely agree with Jose here.